File: indexing_service/services/consumer.js
const Rabbitmq = require('../services/rabbitmq')

class Consumer {
  constructor(queue) {
    this.queue = queue
    this.rabbitmq = new Rabbitmq()
  }

  async subscribe(callback) {
    try {
      this.rabbitmq.channel.assertQueue(this.queue, { durable: true })
      this.rabbitmq.channel.prefetch(1)

      console.log("[Consumer] Waiting for messages in %s. To exit press CTRL+C", this.queue)
      this.rabbitmq.channel.consume(this.queue, async (msg) => {
        await callback(msg)
        this.rabbitmq.channel.ack(msg)
      }, { noAck: false })
    } catch(e) {
      console.log('[Consumer] Error:', e)
      throw e
    }
  }

  async connect() {
    await this.rabbitmq.connect()
  }
}

module.exports = Consumer

File: crawler_service/consumer.py
#!/usr/bin/env python
import pika
import time
import os

class Consumer:
  def __init__(self, queue):
    self.queue = queue
    self.connection = None

  def subscribe(self, callback):
    self.set_connection()
    channel = self.connection.channel()
    channel.queue_declare(queue=self.queue)
    channel.basic_consume(queue=self.queue, on_message_callback=callback, auto_ack=True)

    print('[Consumer] Waiting for messages. To exit press CTRL+C')
    channel.start_consuming()

  def set_connection(self):
    if self.connection is None or not self.connection.is_open:
      rabbitmq_url = os.getenv('RABBITMQ_URL')
      connection_params = pika.URLParameters(rabbitmq_url)
      self.connection = pika.BlockingConnection(connection_params)